### PR TITLE
Test app assigned audio

### PR DESCRIPTION
**Pull request by dautovri**
Assumes an audio controller. 
The key purpose is to verify that a reboot (sudo shutdown -r +1) from the guest doesn’t disrupt the set of assigned audio.
Test checking list of devices before and after reboot

Signed-off-by: Ruslan Dautov <dautov2@gmail.com>

### CODE REVIEW

**Comment by dautovri** [comment]
Added 
```
[device "sound"]
  driver = "ich9-intel-hda"
  bus = "pcie.0"
  addr = "1b.0"
```

passed info via 
```
eden controller edge-node get-config --file $CNF

jq '.deviceIoList += [{ "ptype": "PhyIoAudio", "phylabel": "Audio", "phyaddrs": { "PciLong": "0000:00:1b.0" }, "logicallabel": "Audio", "assigngrp": "Audio", "usage": "PhyIoUsageNone", "usagePolicy": { "freeUplink": false, "fallBackPriority": 0 }, "cbattr": {} }]' < $CNF | \
jq '.apps = (.apps | map(if .displayname == "'$1'" then . + {"adapters":[{"type": "PhyIoAudio", "name": "Audio"}]} else del(.adapters) end))' | \
jq '.apps = (.apps | map(if .displayname == "'$1'" then . + {"purge": {"counter": 1}} else . end))' > $CNF.new
eden controller edge-node set-config --file $CNF.new

```

lspci before config

```
00:00.0 Host bridge: Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
00:02.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:03.0 Communication controller: Red Hat, Inc. Virtio console
00:04.0 Unclassified device [0002]: Red Hat, Inc. Virtio filesystem
00:05.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:1f.0 ISA bridge: Intel Corporation 82801IB (ICH9) LPC Interface Controller (rev 02)
00:1f.2 SATA controller: Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode] (rev 02)
00:1f.3 SMBus: Intel Corporation 82801I (ICH9 Family) SMBus Controller (rev 02)
01:00.0 USB controller: Red Hat, Inc. QEMU XHCI Host Controller (rev 01)
02:00.0 Ethernet controller: Red Hat, Inc. Virtio network device (rev 01)
```

lspci after configs **00:01.0 Audio device** by configs should be on **00:1b.0**

```
00:00.0 Host bridge: Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
00:01.0 Audio device: Intel Corporation 82801I (ICH9 Family) HD Audio Controller (rev 03)
00:02.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:03.0 Communication controller: Red Hat, Inc. Virtio console
00:04.0 Unclassified device [0002]: Red Hat, Inc. Virtio filesystem
00:05.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:1f.0 ISA bridge: Intel Corporation 82801IB (ICH9) LPC Interface Controller (rev 02)
00:1f.2 SATA controller: Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode] (rev 02)
00:1f.3 SMBus: Intel Corporation 82801I (ICH9 Family) SMBus Controller (rev 02)
01:00.0 USB controller: Red Hat, Inc. QEMU XHCI Host Controller (rev 01)
02:00.0 Ethernet controller: Red Hat, Inc. Virtio network device (rev 01)

```
After driver installation 
```

aplay -l
aplay: device_list:274: no soundcards found...

arecord -l
arecord: device_list:274: no soundcards found...
```

File: tests/eclient/qemu+audio.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
```
Comment (giggsoff) [inline tests/eclient/qemu+audio.sh:1]: Please, make this file executable.

File: tests/hardware_reboot/testdata/hardware_audio_reboot.txt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+stdout 'Ubuntu'
+
+# Audio device passthrough setup
+exec -t 20m bash set_audio.sh eclient
+stdout 'msg="Config loaded"'
+
+# Wait for reconfig
+exec -t 20m bash ssh.sh
+stdout 'Ubuntu'
+
```
Comment (giggsoff) [inline tests/hardware_reboot/testdata/hardware_audio_reboot.txt:27]: I think, you should add here `exec -t 2m bash get-lshw.sh {{template "port"}}` to check output before reboot.

File: tests/hardware_reboot/testdata/hardware_audio_reboot.txt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+exec -t 20m bash get-lshw.sh {{template "port"}}
+
+exec -t 5m bash get_reboot_time.sh
+#! cmp stdout reboot_time
+
+# teardown applications
+eden pod delete eclient
+
+test eden.app.test -test.v -timewait 20m - eclient
+
```
Comment (giggsoff) [inline tests/hardware_reboot/testdata/hardware_audio_reboot.txt:47]: Seems, we need to put some comparison and checking of lshw here, without it the test just check app up and down.

**Review by giggsoff — CHANGES_REQUESTED** [review]
Hi, @dautovri 

>lspci after configs 00:01.0 Audio device by configs should be on 00:1b.0

Why do you expect it to be onto `00:1b.0` inside VM in EVE? It is assigned automatically by qemu.

File: tests/eclient/qemu+audio.sh (hunk 1 above)
Comment (giggsoff) [reply tests/eclient/qemu+audio.sh:1]: Probably, the correct variant here is to add `chmod` before the line https://github.com/lf-edge/eden/blob/94cb82fc33cf5d30287f39e537f98e1949997481/tests/eclient/Makefile#L55

**Comment by giggsoff** [comment]
Hi, @dautovri 
I cannot find any drivers in your output loaded for 00:01.0
On Ubuntu VM you can try to run 
```
sudo apt-get install linux-modules-extra-`uname -r`
```
and reboot you VM

**Comment by dautovri** [comment]
@giggsoff Confirmed. On ubuntu VM this command will show a sound card in alsamixer.

```
sudo apt-get install -y linux-modules-extra-`uname -r` alsa-utils
sudo reboot
```
After that 
```
aplay -L
    Discard all samples (playback) or generate zero samples (capture)
surround21:CARD=Intel
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Intel
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Intel
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Intel
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Intel
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Intel
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
```

image

**Comment by zed-sadov** [comment]
All steps need to be automated, including configuring qemu to enable the audio device.

**Comment by sadov** [comment]
@dautovri Seems test in Ubuntu's QEMU is not passing:
```
$ ./eden test tests/hardware_reboot/ -v debug
...
        # Audio device passthrough setup (0.296s)
        > exec -t 20m bash set_audio.sh eclient
        [stdout]
        time="2021-05-14T11:31:01+03:00" level=fatal msg="Cannot unmarshal config: proto:\u00a0(line 324:9): unknown field \"fallBackPriority\""
        [exit status 1]
        FAIL: testdata/hardware_audio_reboot.txt:20: command failure

--- FAIL: TestEdenScripts (0.00s)
    --- FAIL: TestEdenScripts/hardware_audio_reboot (273.19s)
FAIL
...
$ ssh -t -oServerAliveInterval=10 -oConnectTimeout=10 -oStrictHostKeyChecking=no -oPasswordAuthentication=no -i /home/sadov/work/EVE/github/dautovri/eden/dist/tests/eclient/image/cert/id_rsa -p 2223 ubuntu@10.10.10.9
ubuntu@585cd58b-77d1-4489-84e7-369c417f9109:~$ lspci
00:00.0 Host bridge: Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
00:01.0 VGA compatible controller: Device 1234:1111 (rev 02)
00:02.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:03.0 Communication controller: Red Hat, Inc. Virtio console
00:04.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:05.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
00:1f.0 ISA bridge: Intel Corporation 82801IB (ICH9) LPC Interface Controller (rev 02)
00:1f.2 SATA controller: Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode] (rev 02)
00:1f.3 SMBus: Intel Corporation 82801I (ICH9 Family) SMBus Controller (rev 02)
01:00.0 USB controller: Red Hat, Inc. QEMU XHCI Host Controller (rev 01)
02:00.0 SCSI storage controller: Red Hat, Inc. Virtio block device (rev 01)
03:00.0 Ethernet controller: Red Hat, Inc. Virtio network device (rev 01)
ubuntu@585cd58b-77d1-4489-84e7-369c417f9109:~$
```
Are you making any settings for the audio device in QEMU?

Test's log:
[eden_test_hardware_audio_reboot.002.log](https://github.com/lf-edge/eden/files/6477587/eden_test_hardware_audio_reboot.002.log)

**Comment by dautovri** [comment]
@sadov Yes, sure.  QEMU configuration file in PR.  https://github.com/lf-edge/eden/blob/2985d3169359106295c457d3db0ab20585fd078e/tests/eclient/qemu%2Baudio.sh

./qemu+audio.sh
./eden eve stop
./eden eve start
./eden test tests/hardware_reboot/ -v debug

**Comment by sadov** [comment]
@dautovri you need to add this step to the test scenario -- it's a bad idea to run something in tests by hands.

**Comment by mydatascience** [comment]
@sadov @dautovri I believe it's not ready for merge?

**Comment by sadov** [comment]
Yes. As far as I understand, we don't have an automatically working test right now.

**Comment by dautovri** [comment]
`EDEN_TEST_USB_PT=y EDEN_TEST=large ./eden test tests/workflow/ -v debug` 
To check automatically working test.

**Comment by sadov** [comment]
You mean EDEN_TEST_AUDIO_PT?

**Comment by sadov** [comment]
At this time large test failed on `hardware_audio_reboot` test:
```
$ make clean && make build-tests && EDEN_TEST_AUDIO_PT=y EDEN_TEST=large ./eden test tests/workflow/ -v debug
...
        FAIL: ../hardware_reboot/testdata/hardware_audio_reboot.txt:14: command failure

--- FAIL: TestEdenScripts (0.00s)
    --- FAIL: TestEdenScripts/hardware_audio_reboot (1270.10s)
FAIL
```
Full test log:
[eden_test_audio.003.log](https://github.com/lf-edge/eden/files/6493126/eden_test_audio.003.log)

**Comment by dautovri** [comment]
@sadov  Looks like timewait is not enough for app deployment.  In the latest commit, I increased the time and these tests are passing locally.

**Comment by sadov** [comment]
@dautovri local test on Ubuntu 20.04 passed:
```
$ make clean && make build-tests && EDEN_TEST_AUDIO_PT=y EDEN_TEST=large ./eden test tests/workflow/ -v debug
...
--- PASS: TestEdenScripts (0.00s)
    --- PASS: TestEdenScripts/hardware_audio_reboot (981.96s)
PASS
```

**Pull request merged**